nexus_keycode/protocols/full.py

```diff
@@ -104,30 +107,34 @@ def __init__(self, full_id, message_type, body, secret_key, is_factory):
         self.body = body  # shorter body for 'factory' messages
 
         if self.is_factory is True:
-            assert len(body) == 0
+            # assert len(body) == 0
```

Comment:
As you know and according to spec, factory messages *can* have nonzero-length bodies. 

This assert is needed only when the body *is* zero-length. For example, `allow_test` or `display_payg_id` will never have a body. In these functions we pass `body` as an empty string, `""`. Could we update this code to:
1. detect if `body` is `""`, and if so, use this assert and set `body_int` to 0
2. update the docstring of this `init` function so that `body` is String type (this is what it is now)

So at the end of this,
- *all* factory messages should have `full_id` == 0
- `len(body) == 0` and `self.body_int` == 0 iff `body == ""`

If we make these changes, then I think all messages will work properly: MAC will be computed for activation messages, and factory messages with or without body digits. They will not be computed for passthrough commands due to the `if` you added below.